### DISCUSSION TITLE
nextjs with directus/sdk-js

### DISCUSSION THREAD

**Discussion by jiangyh1024** [directus/directus#3442] [Archive]
![image](https://user-images.githubusercontent.com/34930870/102463765-8076d980-4086-11eb-81fb-d812bb2d99e5.png)

**Comment by jiangyh1024** [comment]
maybe caused by 
![image](https://user-images.githubusercontent.com/34930870/102464086-e4999d80-4086-11eb-889b-be90ba792434.png)
 ?

**Reply by TorbenHammes** [reply]
I have the same issue

**Comment by rijkvanzanten** [comment]
@jiangyh1024 `next` should handle optional chaining [starting with v9.1.7](https://nextjs.org/blog/next-9-1-7). What bundler are you using for your node modules? Make sure it's up to date enough to handle Optional Chaining in JS.

(@nickrum Seems like Next.js is importing from the `module` path instead of the `main` path. Is the `main` export (the CJS one) babel-ed down?)

**Reply by jiangyh1024** [reply]
the project is created by  npx create-next-app projectname. It uses webpack as the bundler.

Another project created by npx create-react-app projectname works well. But I want to use nextjs

**Reply by nickrum** [reply]
It seems like Next.js (or rather webpack 4, which is used by next under the hood) handles imported libraries and code written by the user differently. When using Optional Chaining in a freshly created Next.js project, it works perfectly fine.

(@rijkvanzanten Yep, webpack 4 prefers to use es modules for better tree-shaking and only falls back to using the `main` export if no esm version is provided. The `main` and the `module` export both compile to esnext. The idea behind that was to provide the smallest bundle possible and to let the bundler importing either from `main` or from `module` do the babel-ing based on the users needs.)

**Reply by rijkvanzanten** [reply]
> It seems like Next.js (or rather webpack 4, which is used by next under the hood) handles imported libraries and code written by the user differently. When using Optional Chaining in a freshly created Next.js project, it works perfectly fine.

This feels weird.. I'm almost wondering if we should open an issue on the Next repo about this.

**Reply by nickrum** [reply]
~Sounds like a good idea. As other libraries do something similar, I don't think this is an issue on our side.~

**Reply by nickrum** [reply]
Actually, scratch that. Webpack's babel-loader recommends not to compile `node_modules`, so this is somewhat expected. Version 5 of webpack updates the internal parser to also support more recent JS syntax (including Optional Chaining). So there is not much we can do except to wait until webpack 5 is more widely adopted and in the meantime fall back to es2019 or earlier.

**Reply by danbas90** [reply]
Nextjs supports webpack 5 by using yarn resolutions if that helps

https://nextjs.org/blog/next-9-5#webpack-5-support-beta

**Reply by nickrum** [reply]
Thanks @danbas90, I saw that as well. Unfortunately, I don't think it would be a good idea to require the use of yarn just to make the directus sdk work.

**Reply by WoLfulus** [reply]
@rijkvanzanten can we target a lower ES version on sdk?

**Reply by rijkvanzanten** [reply]
@WoLfulus yup, too bad that the node ecosystem is getting so fragmented, but it is what it is

**Comment by Jonesus** [comment]
I managed to work around this issue for now with `next-transpile-modules` (https://github.com/martpie/next-transpile-modules), following their readme and creating a `next.config.js` (for my configuration) like so:

```
const withTM = require('next-transpile-modules')(['@directus/sdk-js']);

module.exports = withTM();
```

Working for me at least on `next@10.0.4`, `@directus/sdk-js@9.0.0-rc.27` and `next-transpile-modules@6.0.0` but I don't see a reason why this wouldn't work with different `next` or `@directus/sdk-js` versions
